File: lib/eval/scope.ml
type obj
   = Null
   | Int of int
   | Function of {
      name: string;
      (* args: Ast.expr list; *)
      body: Ast.expr;
   }
   | Builtin of (obj list -> obj)
;;

type scope = (string, obj) Hashtbl.t;;
type scope_stack = scope list;;

let _list_varnames (stack:scope_stack) =
   let _ = print_endline ("registered: " ^
      (Int.to_string (List.fold_left 
         (fun sum scope -> sum + Hashtbl.length scope)
         0
         stack))) in
   let _ = List.map (fun scope -> Hashtbl.iter (fun name _val -> print_endline name) scope) stack
   in ()

let create_scope_stack =
   [Hashtbl.create 10]
;;

let add_scope stack =
   Hashtbl.create 0 :: stack
;;

let pop_scope (stack:scope_stack) =
   match stack with
   | _ :: rest -> rest
   | _ -> raise (Failure "bug: empty scope stack")
;;

let define_local (stack:scope_stack) name value =
   Hashtbl.replace (List.nth stack 0) name value
;;

let find (stack:scope_stack) name =
   List.find_map (fun scope -> Hashtbl.find_opt scope name) stack
;;

File: src/eval/scope.rs

use super::object::Object;
use std::collections::HashMap;

struct Scope<'a>(HashMap<String, Object<'a>>);
pub struct ScopeStack<'a>{
    scopes: Vec<Scope<'a>>,
}

impl Scope<'_> {
    fn new<'a>() -> Scope<'a> {
        Scope(HashMap::new())
    }
}

impl<'scope> ScopeStack<'scope> {
    pub fn new<'a>() -> ScopeStack<'a> {
        ScopeStack { scopes: vec![Scope::new()] }
    }

    pub fn find<'this, 'obj: 'this>(&'this self, name: &str) -> Option<Object<'obj>> {
        self
            .scopes
            .iter()
            .rev()
            .find_map(|scope| scope.0.get(name).map(|obj| obj.clone()))
    }

    pub fn declare(&mut self, name: String, val: Object) {
        self
            .scopes
            .last_mut()
            .expect("bug: empty scope stack")
            .0
            .insert(name, val);
    }

    pub fn add_scope(&mut self) {
        self.scopes.push(Scope::new())
    }
}
